### PR TITLE
Model refactoring

### PR DESCRIPTION
**Pull request by Rasmusafj**
- Added all the final models and correct hashes after repackaging them all.
- Refactored model.py

### CODE REVIEW

File: danspeech/deepspeech/model.py [hunk 1]
```diff
@@ -592,22 +535,142 @@ def forward(self, x, is_first, is_last):
...
+                seq_len = ((seq_len + 2 * m.padding[1] - m.dilation[1] * (m.kernel_size[1] - 1) - 1) / m.stride[1] + 1)
+        return seq_len.int()
+
     @classmethod
     def load_model(cls, path):
+        """
+        If you do not need meta data from package, simply use this method to load the model.
+
+        :param path: Path to .pth package for the model.
+        :param conv_layers: Number of Conv layers. Will be removed after repackaging models.
```
Comment (zer0x42) [inline danspeech/deepspeech/model.py]: This isn't the case, is it? We need to differentiate between different number of convs?

File: danspeech/deepspeech/model.py [hunk 2]
```diff
@@ -592,22 +535,142 @@ def forward(self, x, is_first, is_last):
...
+            x.flatten_parameters()
+        return model
+
+    @classmethod
+    def load_model_package(cls, package):
+        """
+        If you need aditional information from package, then use this loading method instead.
+
+        :param package: Package holding the model,
+        :param conv_layers: conv_layers: Number of Conv layers. Will be removed after repackaging models.
```
Comment (zer0x42) [inline danspeech/deepspeech/model.py]: Same as above

File: danspeech/deepspeech/model.py [hunk 3]
```diff
@@ -592,22 +535,142 @@ def forward(self, x, is_first, is_last):
...
         return model
 
+    @staticmethod
+    def serialize(model, optimizer=None, epoch=None, iteration=None, loss_results=None,
+                  cer_results_corr=None, wer_results_corr=None, cer_results_ncorr=None,
+                  wer_results_ncorr=None, avg_loss=None, meta=None, distributed=False):
+        """
+
+        # TODO: Refactor but find out what we want to serialize.
+        # When is this method needed. Should it really lie in training repo? or ?
```
Comment (zer0x42) [inline danspeech/deepspeech/model.py]: Should be in training repo imho

File: danspeech/deepspeech/model.py [hunk 4]
```diff
@@ -592,22 +535,142 @@ def forward(self, x, is_first, is_last):
...
+    def get_param_size(model):
+        params = 0
+        for p in model.parameters():
+            tmp = 1
+            for x in p.size():
+                tmp *= x
+            params += tmp
+        return params
 
 if __name__ == '__main__':
```
Comment (zer0x42) [inline danspeech/deepspeech/model.py]: There's not much point in having a run example in here when we document the package imo.

File: danspeech/pretrained_models/baseline_model.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
+from danspeech.deepspeech.model import DeepSpeech
+from danspeech.utils.data_utils import get_model
+
+MODEL_PACKAGE = "toDo"
+
+# ToDO: Create
+def Baseline(cache_dir=None):
+    """
+    Instantiates the most complex DanSpeech model with a lot of parameters
```
Comment (zer0x42) [inline danspeech/pretrained_models/baseline_model.py]: Wrong docstring

File: danspeech/pretrained_models/custom_model.py [hunk 6]
```diff
@@ -0,0 +1,13 @@
+
+from danspeech.deepspeech.model import DeepSpeech
+
+# ToDO: Create
+def CustomModel(model_path):
+    """
+    Instantiates the most complex DanSpeech model with a lot of parameters
```
Comment (zer0x42) [inline danspeech/pretrained_models/custom_model.py]: wrong docstring

File: danspeech/pretrained_models/folketinget_model.py [hunk 7]
```diff
@@ -0,0 +1,16 @@
+from danspeech.deepspeech.model import DeepSpeech
+from danspeech.utils.data_utils import get_model
+
+MODEL_PACKAGE = 'https://github.com/Rasmusafj/models_development/raw/master/400units.pth'
```
Comment (zer0x42) [inline danspeech/pretrained_models/folketinget_model.py]: Dangerous to have a wrong package link, should be toDo I guess?

File: danspeech/pretrained_models/streaming_model_CPU.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+from danspeech.deepspeech.model import DeepSpeech
+from danspeech.utils.data_utils import get_model
+
+MODEL_PACKAGE = "toDo"
+
+# ToDO: Create
+def CPUStreamingRNN(cache_dir=None):
+    """
+    Instantiates the most complex DanSpeech model with a lot of parameters
```
Comment (zer0x42) [inline danspeech/pretrained_models/streaming_model_CPU.py]: Docstring

File: danspeech/pretrained_models/streaming_model_GPU.py [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+from danspeech.deepspeech.model import DeepSpeech
+from danspeech.utils.data_utils import get_model
+
+MODEL_PACKAGE = "toDo"
+
+# ToDO: Create
+def GPUStreamingRNN(cache_dir=None):
+    """
+    Instantiates the most complex DanSpeech model with a lot of parameters
```
Comment (zer0x42) [inline danspeech/pretrained_models/streaming_model_GPU.py]: docstring

File: danspeech/deepspeech/model.py (hunk 1 above)
Comment (Rasmusafj) [reply danspeech/deepspeech/model.py]: Fixed after repackageing. Forgot about comment! Nice spot. :)

File: danspeech/deepspeech/model.py (hunk 2 above)
Comment (Rasmusafj) [reply danspeech/deepspeech/model.py]: Fixed after repackageing. Forgot about comment! Nice spot. :)

File: danspeech/pretrained_models/streaming_model_GPU.py [hunk 10]
```diff
@@ -0,0 +1,17 @@
+import torch
```
Comment (zer0x42) [inline danspeech/pretrained_models/streaming_model_GPU.py]: torch has been disabled, in non streaming models but not here - is that intentional?

File: danspeech/deepspeech/model.py (hunk 3 above)
Comment (Rasmusafj) [reply danspeech/deepspeech/model.py]: Remove this method in this PR then?

File: danspeech/deepspeech/model.py (hunk 4 above)
Comment (Rasmusafj) [reply danspeech/deepspeech/model.py]: Agree! Removed.

**Review by zer0x42 — COMMENTED** [review]
The model scripts need docstring updates, and a decision on whether or not to import torch (torch might be required for streaming, but not for bi-directional? I just don't see why)

File: danspeech/pretrained_models/custom_model.py (hunk 6 above)
Comment (Rasmusafj) [reply danspeech/pretrained_models/custom_model.py]: Updated all docstrings.

File: danspeech/pretrained_models/folketinget_model.py (hunk 7 above)
Comment (Rasmusafj) [reply danspeech/pretrained_models/folketinget_model.py]: Yep! Changed to todo

File: danspeech/deepspeech/model.py (hunk 3 above)
Comment (Rasmusafj) [reply danspeech/deepspeech/model.py]: Removed it!

**Pull request merged**